Avoid assignment checking for signature polymorphic methods

SirYwell opened the pull request:
Hi, this is a very local fix of how [signature polymorphic](https://docs.oracle.com/javase/specs/jls/se25/html/jls-15.html#jls-15.12.3) method invocations are handled during type checking.

Somehow, invoking such a method with only one argument succeeded before if the argument is a reference type. For some reason, e.g. String seems to be considered a subtype of Object[]. This is not the case for primitive types, leading to a crash.

As the signature of an invocation of a signature polymorphic method depends only on the static types at call site, it doesn't really make sense to do type checking there.

There might be better ways to handle signature polymorphic methods in general, but that's probably also a bit more invasive. This change mainly targets the crash I encountered.

wmdietl commented:
Hi @SirYwell, thanks for the pull request and fixing a crash!

The `misc` test currently fails because your commit shows up with two different authors. Could you re-write your commit to just keep your real name?
 
````
--- a/docs/manual/contributors.tex
+++ b/docs/manual/contributors.tex
@@ -48,6 +48,7 @@ Google Inc.\ (via @wmdietlGC),
 Haaris Ahmed,
 Haifeng Shi,
 Hamed Taghani,
+Hannes Greule,
 Heath Borders,
 Ivory Wang,
 Jakub Vr\'ana,
@@ -136,6 +137,7 @@ Shinya Yoshida,
 Shubham Raj,
 Sidney Monteiro,
 Simon Gerst,
+SirYwell,
 Stefan Heule,
 Steph Dietzel,
 Stephan Schroevers,
+ set +x
docs/manual/contributors.tex is not up to date.
If the above suggestion is appropriate, run: make -C docs/manual contributors.tex
If the suggestion contains a username rather than a human name, then do all the following:
  * Update your git configuration by running:  git config --global user.name "YOURFULLNAME"
  * Add your name to your GitHub account profile at https://github.com/settings/profile
  * Make a pull request to add your GitHub ID to
    https://github.com/eisop-plume-lib/git-scripts/blob/master/git-authors.sed
    and remake contributors.tex after that pull request is merged.
Error: Process completed with exit code 1.
````

Just to document the crash you note, the test file without the fix produces this stack trace, multiple times:

````
error: DefaultTypeHierarchy: unexpected combination:  type1: [INT class org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedPrimitiveType] int  type2: [ARRAY class org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedArrayType] Object []
    visitHistory = VisitHistory( {IPair(Object, Object)=[@org.checkerframework.checker.nullness.qual.UnknownKeyFor], IPair(Object, Object)=[@org.checkerframework.checker.nullness.qual.UnknownKeyFor], IPair(MethodHandle, MethodHandle)=[@org.checkerframework.checker.nullness.qual.UnknownKeyFor]} )
  ; The Checker Framework crashed.  Please report the crash.  Version: Checker Framework 3.49.5-eisop1-SNAPSHOT, branch master, 2025-07-24, commit 2a3c94c. 
  Checker: class org.checkerframework.checker.nullness.KeyForSubchecker
  Visitor: class org.checkerframework.common.basetype.BaseTypeVisitor
  Compilation unit: framework/tests/all-systems/Issue6078.java
  Last visited tree at line 13 column 9:
  	methodHandle.invoke(1);
  Exception: java.lang.Throwable; java.lang.Throwable
  	at org.checkerframework.javacutil.BugInCF.<init>(BugInCF.java:27)
  	at org.checkerframework.framework.type.visitor.AtmComboVisitor.defaultAction(AtmComboVisitor.java:66)
  	at org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visitPrimitive_Array(AbstractAtmComboVisitor.java:391)
  	at org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:541)
  	at org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:265)
  	at org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:141)
  	at org.checkerframework.common.basetype.BaseTypeVisitor.commonAssignmentCheck(BaseTypeVisitor.java:3527)
  	at org.checkerframework.common.basetype.BaseTypeVisitor.commonAssignmentCheck(BaseTypeVisitor.java:3501)
  	at org.checkerframework.common.basetype.BaseTypeVisitor.checkArguments(BaseTypeVisitor.java:4138)
  	at org.checkerframework.common.basetype.BaseTypeVisitor.visitMethodInvocation(BaseTypeVisitor.java:2173)
  	at org.checkerframework.common.basetype.BaseTypeVisitor.visitMethodInvocation(BaseTypeVisitor.java:197)
  	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1832)
  	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:92)
  	at org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:97)
  	at org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:483)
  	at org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:197)
  	at jdk.compiler/com.sun.source.util.TreeScanner.visitExpressionStatement(TreeScanner.java:503)
  	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCExpressionStatement.accept(JCTree.java:1611)
  	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:92)
````

SirYwell commented:
Thanks for taking a look @wmdietl, I think I successfully changed the name.

wmdietl commented on javacutil/src/main/java/org/checkerframework/javacutil/TreeUtils.java:
```diff
@@ -2702,23 +2701,20 @@ public static boolean isVarargsCall(MethodInvocationTree invok) {
 
-        return false;
+    /**
+     * Returns true if the given method invocation targets a signature polymorphic method. See <a
```
I would have also written `signature-polymorphic method`, but the JLS consistently writes it without dash, so let's go with that.

wmdietl reviewed (approved):
Thanks again for the fix!
I've added you to the contributors file and rewrote some comments. Please take a look and let me know whether things look good.

SirYwell commented:
> I've added you to the contributors file and rewrote some comments. Please take a look and let me know whether things look good.

The comments look good to me, thanks!
